### PR TITLE
Additional autograd unit tests for Python UDFs.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28824 Additional autograd unit tests for Python UDFs.**

1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

### CODE REVIEW

File: torch/csrc/Exceptions.h [hunk 1]
```diff
@@ -69,6 +69,23 @@ struct python_error : public std::exception {
     }
   }
 
+  virtual const char* what() const noexcept override {
+    // Retrieve the error message from the value.
+    if (value != nullptr && value != Py_None) {
+      AutoGIL gil;
+      Py_XINCREF(value);
```
Comment (ezyang) [inline torch/csrc/Exceptions.h]: Manual refcounting, blegh

File: torch/csrc/autograd/python_engine.h [hunk 2]
```diff
@@ -21,6 +21,10 @@ struct PythonEngine : public Engine {
       bool keep_graph,
       bool create_graph,
       const edge_list& outputs = {}) override;
+
+  variable_list execute_with_graph_task(
+      std::shared_ptr<GraphTask> graph_task,
+      std::shared_ptr<Node> graph_root);
```
Comment (ezyang) [inline torch/csrc/autograd/python_engine.h]: Mark this `override` right?

File: torch/csrc/autograd/python_engine.cpp [hunk 3]
```diff
@@ -70,6 +70,16 @@ variable_list PythonEngine::execute(
   }
 }
 
+variable_list PythonEngine::execute_with_graph_task(
+    std::shared_ptr<GraphTask> graph_task,
+    std::shared_ptr<Node> graph_root) {
+  try {
+    return Engine::execute_with_graph_task(graph_task, graph_root);
+  } catch (python_error& e) {
+    e.restore();
```
Comment (ezyang) [inline torch/csrc/autograd/python_engine.cpp]: This looks pretty weird to me. What it kind of looks like is that you reraised a `python_error` from another thread but forgot to restore the Python error context at the time you threw the error, so now you have to "fix" it here. Wouldn't it be better to just handle it properly at the rethrowing site?

File: torch/csrc/Exceptions.h (hunk 1 above)
Comment (ezyang) [reply torch/csrc/Exceptions.h]: I don't see a paired decref, nor do I see a steal on value, so it looks to me like you're leaking `value`?

File: torch/csrc/Exceptions.h [hunk 4]
```diff
@@ -69,6 +69,23 @@ struct python_error : public std::exception {
...
+      PyObject* pyStr = PyObject_Str(value);
+      if (pyStr != Py_None && pyStr != nullptr) {
```
Comment (ezyang) [inline torch/csrc/Exceptions.h]: Even if `pyStr` returns a `Py_None` you are obligated to decref it (yes, `Py_None` is refcounted)

Comment (ezyang) [reply torch/csrc/Exceptions.h]: If it returns `nullptr` you are obligated to propagate the Python exception, or clear it

File: torch/csrc/Exceptions.h [hunk 5]
```diff
@@ -69,6 +69,23 @@ struct python_error : public std::exception {
...
+        PyObject* str = PyUnicode_AsEncodedString(pyStr, "utf-8", "strict");
+        const auto& ret = PyBytes_AS_STRING(str);
```
Comment (ezyang) [inline torch/csrc/Exceptions.h]: You can't do this. You need to keep `str` live for the returned `char*` to be valid, but there's no way in this API to do so. Your `char*` goes immediately to dead data after you do this

**Review by ezyang — CHANGES_REQUESTED** [review]
Major errors

File: test/dist_autograd_test.py [hunk 6]
```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
...
+    def test_backward_python_udf_error(self):
+        t1 = torch.rand((3, 3), requires_grad=True)
+        t2 = torch.rand((3, 3), requires_grad=True)
+        with dist_autograd.context() as context_id:
+            loss = rpc.rpc_sync('worker{}'.format(self._next_rank()), DistAutogradTest._nested_rpc_call_backward_error, args=(t1, t2, self._next_rank()))
+            with self.assertRaisesRegex(RuntimeError, 'Simulate error on backward pass'):
+                dist_autograd.backward([loss.sum()])
+
+
+    _backward_done = False
```
Comment (mrshenli) [inline test/dist_autograd_test.py]: Other state vars are out of the test class. Any reason for putting this here?

File: test/dist_autograd_test.py [hunk 7]
```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
...
+
+    def _set_backward_done():
+        DistAutogradTest._backward_done = True
+
+    def _wait_backward_done():
+        while not DistAutogradTest._backward_done:
+            time.sleep(0.1)
+            pass
```
Comment (mrshenli) [inline test/dist_autograd_test.py]: why do we need this `pass` here?

File: test/dist_autograd_test.py [hunk 8]
```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
...
+        res = rpc.rpc_sync('worker{}'.format(dst), DistAutogradTest._python_udf_with_backward_error, args=(t1, t2))
+        return torch.chain_matmul(t1, t2, res)
+
+
+    @dist_init
+    def test_backward_python_udf_error(self):
+        t1 = torch.rand((3, 3), requires_grad=True)
+        t2 = torch.rand((3, 3), requires_grad=True)
+        with dist_autograd.context() as context_id:
+            loss = rpc.rpc_sync('worker{}'.format(self._next_rank()), DistAutogradTest._nested_rpc_call_backward_error, args=(t1, t2, self._next_rank()))
```
Comment (mrshenli) [inline test/dist_autograd_test.py]: There are several super long lines. Let's break them into shorter ones.

File: test/dist_autograd_test.py [hunk 9]
```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
...
+            # Wait for all RPCs to be done.
+            dist.barrier()
+
+            # Kill rank 2 (last hop of nested rpc) and verify rank 0 receives an error.
+            if self.rank == 2:
+                return
+
+            if self.rank == 0:
+                # Wait a bit for rank 2 to die.
+                time.sleep(5)
```
Comment (mrshenli) [inline test/dist_autograd_test.py]: This depends on physical time to sync processes, which might potentially make this test flaky. Is there a way to implement the ordering explicitly?

File: torch/csrc/autograd/python_engine.cpp (hunk 3 above)
Comment (pritamdamania87) [reply torch/csrc/autograd/python_engine.cpp]: There is a case where this method is called directly from `DistEngine` in that case, we would like to restore the error. I'll change this to only do `e.restore` if the error is not set already.

**Comment by zhaojuanmao** [comment]
the tests look good to me, wondering whether it could be mixed with refs as well for these tests

**Review by ezyang — APPROVED** [review]
I didn't review the newly added tests carefully, please get someone else to review them

**Comment by pritamdamania87** [comment]
@zhaojuanmao I believe Shen added a bunch of tests for RRefs in https://github.com/pytorch/pytorch/pull/28656. Do you have any additional tests in mind?

File: test/dist_autograd_test.py [hunk 10]
```diff
@@ -103,6 +102,22 @@ def _all_contexts_cleaned_up(timeout_seconds=10):
     success = len(context_id_to_raised) == len(known_context_ids) and all(context_id_to_raised.values())
     return success
 
+def noop():
+    pass
+
+def wait_until_node_failure(rank):
+    '''
+        Loops until an RPC to the given rank doesn't fail. This is used to
```
Comment (mrshenli) [inline test/dist_autograd_test.py]: you mean loops until ... DOES fail?

**Review by mrshenli — APPROVED** [review]
Lint failures on `python_engine.cpp` does not seem relevant, but please double check. Tests look good to me.

**Comment by ezyang** [comment]
This diff caused C++ docs check to fail on master:

```
+ grep -v 'warning: Included by graph for '\''types.h'\'' not generated, too many nodes' doxygen-log.txt
+ set -e
+ mv temp.txt doxygen-log.txt
++ grep warning: doxygen-log.txt
++ wc -l
+ warnings=1
+ echo 'Treating all remaining warnings as errors'
+ [[ 1 -ne 0 ]]
+ echo 'Failing Doxygen test because the following warnings were treated fatally:'
+ cat doxygen-log.txt
Treating all remaining warnings as errors
Failing Doxygen test because the following warnings were treated fatally:
warning: Included by graph for 'Exception.h' not generated, too many nodes. Consider increasing DOT_GRAPH_MAX_NODES.
+ echo 'Please fix these warnings.  To run this test locally, use docs/cpp/source/check-doxygen.sh'
Please fix these warnings.  To run this test locally, use docs/cpp/source/check-doxygen.sh
+ rm -f doxygen-log.txt original-doxygen-log.txt
+ exit 1
```

**Pull request closed without merging**